domain/credential/service/service.go

```diff
@@ -50,31 +57,90 @@ type State interface {
 		getWatcher func(string, string, changestream.ChangeType) (watcher.NotifyWatcher, error),
 		name, cloudName, owner string,
 	) (watcher.NotifyWatcher, error)
+
+	// ModelsUsingCloudCredential returns a map of uuid->name for models which use the credential.
+	ModelsUsingCloudCredential(ctx context.Context, id credential.ID) (map[model.UUID]string, error)
+
+	// GetCloud returns the cloud with the specified name.
+	GetCloud(ctx context.Context, name string) (cloud.Cloud, error)
 }
 
+// ValidationContextGetter returns the artefacts for a specified model, used to make credential validation calls.
+type ValidationContextGetter func(modelUUID model.UUID) (credentialcommon.CredentialValidationContext, error)
+
 // Logger facilitates emitting log messages.
 type Logger interface {
-	Debugf(string, ...interface{})
+	Warningf(string, ...interface{})
 }
 
 // Service provides the API for working with external controllers.
 type Service struct {
 	st             State
 	watcherFactory WatcherFactory
+	logger         Logger
+
+	// These are set via options after the service is created.
+
+	validationContextGetter ValidationContextGetter
+	validator               credentialcommon.CredentialValidator
+
+	// TODO(wallyworld) - remove when models are out of mongo
+	legacyUpdater func(tag names.CloudCredentialTag) error
+	legacyRemover func(tag names.CloudCredentialTag) error
 }
 
 // NewService returns a new service reference wrapping the input state.
-func NewService(st State, watcherFactory WatcherFactory) *Service {
+func NewService(st State, watcherFactory WatcherFactory, logger Logger) *Service {
 	return &Service{
 		st:             st,
 		watcherFactory: watcherFactory,
+		logger:         logger,
+		validator:      noopValidator{},
 	}
 }
 
-// CloudCredential represents a credential and the cloud it belongs to.
-type CloudCredential struct {
-	Credential cloud.Credential
-	CloudName  string
+// noopValidator always validates any credential.
+// TODO(wallyworld) - this is a placeholder.
+type noopValidator struct {
+}
+
+func (noopValidator) Validate(
+	_ credentialcommon.CredentialValidationContext,
+	_ credential.ID,
+	_ *cloud.Credential,
+	_ bool,
+) (machineErrors []error, err error) {
```

Comment:
Disagree - they were put there to aid readability since "[]error" and "error" is confusing.
In any case, this noopValidator is removed in the next PR - it's just a placeholder.

Reply:
Yeah but what is a machineErrors ? That's what confused me.

Reply:
It's documented in the proper validator. This is just a placeholder just for this PR. I'll remove them simply because the whole thing is deleted next PR anyway.